<a href="https://colab.research.google.com/github/W-Srikongpan/CKD/blob/main/step2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Import

In [1]:
from google.colab import drive
import numpy as np
import pandas as pd
drive.mount('/content/drive')

Mounted at /content/drive


#Data Lab

In [2]:
lab = '/content/drive/My Drive/ProjectCKD/df_lab.csv'
df_lab = pd.read_csv(lab)
df_lab.head(5)

LabNo  PatientUID  PatientVisitUID  \
0      1        2299          2024049   
1      1        2299          2024049   
2      1        2299          2024049   
3      1        2299          2024049   
4      1        2299          2024049   

                                    Location RequestItemName  \
0              เวชศาสตร์นิวเคลียร์ (OPD)-D15      BMD LUMBAR   
1  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร  ALBUMIN, 32403   
2  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   Alkaline Phos   
3  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร      ALT, 32311   
4  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร      AST, 32310   

     ResultItemName                    CWhen ResultDttm ResultValue Unit  \
0  Formatted Result  2015-02-19 16:16:51.423        NaN          FT  NaN   
1    ALBUMIN, 32403  2015-02-19 10:32:20.907        NaN         4.1  NaN   
2        ALP, 32309  2015-02-19 10:32:21.610        NaN          44  NaN   
3        ALT, 32311  2015-02-19 10:32:21.390        NaN          17  NaN   
4        AST, 32310  2015-02-19 10:32:21.500        NaN          23  NaN   

  ReferenceRange          Lab  
0            NaN          NaN  
1      3.8 - 5.4  L11_Albumin  
2       42 - 121          NaN  
3         4 - 36          NaN  
4        12 - 32          NaN

#Data Laboratory

In [3]:
Laboratory = '/content/drive/My Drive/ProjectCKD/df_Laboratory.csv'
df_Laboratory = pd.read_csv(Laboratory)
df_Laboratory.head()

True laboratory item  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0       L1_eGFR            eGFR           1      1587.0         NaN   
1        L2_BUN             BUN           3       129.0       185.0   
2         L3_Cr              Cr           0       103.0         NaN   
3     L4_Sodium          Sodium           5       115.0         NaN   
4  L5_Potassium       Potassium           2       112.0         NaN   

   Unnamed: 7  Unnamed: 8  Unnamed: 9  
0         NaN         NaN         NaN  
1       312.0         NaN         NaN  
2         NaN         NaN         NaN  
3         NaN         NaN         NaN  
4         NaN         NaN         NaN

#Create new df is df_lab2

In [4]:
df_lab2 = {}
df_lab2 = pd.DataFrame(df_lab2)
df_lab2['PatientUID'] = df_lab.PatientUID
df_lab2['PatientVisitUID'] = df_lab.PatientVisitUID
df_lab2['CWhenfirst'] = df_lab.CWhen
df_lab2['ResultValue'] = df_lab.ResultValue
df_lab2['Lab'] = df_lab.Lab
df_lab2.head(10)

PatientUID  PatientVisitUID               CWhenfirst ResultValue  \
0        2299          2024049  2015-02-19 16:16:51.423          FT   
1        2299          2024049  2015-02-19 10:32:20.907         4.1   
2        2299          2024049  2015-02-19 10:32:21.610          44   
3        2299          2024049  2015-02-19 10:32:21.390          17   
4        2299          2024049  2015-02-19 10:32:21.500          23   
5        2299          2024049  2015-02-19 10:32:20.140         9.5   
6        2299          2024049  2015-02-19 10:32:21.717         9.4   
7        2299          2024049  2015-02-19 10:32:20.343         190   
8        2299          2024049  2015-02-19 10:32:20.233         0.9   
9        2299          2024049  2015-02-19 10:32:21.280         0.1   

               Lab  
0              NaN  
1      L11_Albumin  
2              NaN  
3              NaN  
4              NaN  
5           L2_BUN  
6       L9_Calcium  
7  L21_Cholesterol  
8            L3_Cr  
9              NaN

In [6]:
df_lab2.to_csv('/content/drive/My Drive/ProjectCKD/df_eGFR.csv', index=False)

#Create new df is df_eGFR

In [7]:
df_eGFR = df_lab2[df_lab2.Lab == 'L1_eGFR']
df_eGFR.head(10)

PatientUID  PatientVisitUID               CWhenfirst ResultValue      Lab
248        2299          3386255  2016-08-15 09:51:54.913        71.7  L1_eGFR
255        2299          3640924  2016-11-11 09:40:06.053        66.8  L1_eGFR
271        2299          4020970  2017-03-31 10:18:54.363       67.20  L1_eGFR
286        2299          4286612  2017-07-05 10:40:40.763       69.20  L1_eGFR
331        2299          4546471  2017-10-03 09:19:54.710       63.60  L1_eGFR
343        2299          4801938  2018-01-04 09:03:28.847       58.40  L1_eGFR
370        2299          5051781  2018-03-29 09:10:51.133       62.40  L1_eGFR
378        2299          5296557  2018-06-21 08:55:47.877       55.00  L1_eGFR
391        2299          5377983  2018-07-18 10:28:12.007       64.00  L1_eGFR
442        2299          5544604  2018-09-12 09:13:36.107       62.40  L1_eGFR

Drop row eGFR is NaN

In [8]:
df_eGFR['ResultValue'] = pd.to_numeric(df_eGFR['ResultValue'],errors= 'coerce')
df_eGFR['ResultValue'].isna().sum()

<ipython-input-8-5155dd8b23f1>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_eGFR['ResultValue'] = pd.to_numeric(df_eGFR['ResultValue'],errors= 'coerce')


64

df_eGFR has NaN 64 rows

In [9]:
df_eGFR[df_eGFR['ResultValue'].isna()]

PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
1330622       119514          3571540  2016-10-18 20:48:28.817          NaN   
4453496       563983          3804017  2017-01-12 13:04:14.380          NaN   
4471358       566187          8235036  2021-03-19 09:09:07.297          NaN   
5232553       690624          7797789  2020-10-21 16:16:58.290          NaN   
6486667       871360          4278938  2017-07-03 15:04:50.793          NaN   
...              ...              ...                      ...          ...   
20999454      464537          4882028  2018-02-28 09:50:42.277          NaN   
21191192      738697          5438058  2018-09-07 08:31:36.957          NaN   
21215351     1168494          9701496  2022-09-01 07:56:11.627          NaN   
21260028       22211          3392590  2016-10-02 08:52:02.277          NaN   
21272957     2261515          3927229  2017-03-31 08:57:10.647          NaN   

              Lab  
1330622   L1_eGFR  
4453496   L1_eGFR  
4471358   L1_eGFR  
5232553   L1_eGFR  
6486667   L1_eGFR  
...           ...  
20999454  L1_eGFR  
21191192  L1_eGFR  
21215351  L1_eGFR  
21260028  L1_eGFR  
21272957  L1_eGFR  

[64 rows x 5 columns]

In [10]:
df_eGFR = df_eGFR.dropna(subset=['ResultValue'])
df_eGFR

PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
248             2299          3386255  2016-08-15 09:51:54.913         71.7   
255             2299          3640924  2016-11-11 09:40:06.053         66.8   
271             2299          4020970  2017-03-31 10:18:54.363         67.2   
286             2299          4286612  2017-07-05 10:40:40.763         69.2   
331             2299          4546471  2017-10-03 09:19:54.710         63.6   
...              ...              ...                      ...          ...   
21628049      925649          3617725  2017-04-17 08:54:35.740         77.6   
21628297      925649          3617725  2017-04-19 08:41:45.060         72.0   
21628538      925649          3617725  2017-04-21 08:47:03.167         62.9   
21628775      925649          3617725  2017-04-24 08:59:53.310         59.6   
21629006      925649          3617725  2017-04-26 11:02:13.990         60.1   

              Lab  
248       L1_eGFR  
255       L1_eGFR  
271       L1_eGFR  
286       L1_eGFR  
331       L1_eGFR  
...           ...  
21628049  L1_eGFR  
21628297  L1_eGFR  
21628538  L1_eGFR  
21628775  L1_eGFR  
21629006  L1_eGFR  

[518647 rows x 5 columns]

    First df_eGFR has 518711 rows
    and df_eGFR Nan has 64 rows,
    Therefor df_eGFR has 518647 rows  
    



#Check CKD stage 5


    group_eGFR1 is df_eGFR groupby and first row
    group_eGFR2 is df_eGFR groupby and last row

In [11]:
df_eGFR = df_eGFR.sort_values('CWhenfirst')
group_eGFR1 = df_eGFR.groupby('PatientUID').first().reset_index()
group_eGFR2 = df_eGFR.groupby('PatientUID').last().reset_index()
group_eGFR1.head(10)

PatientUID  PatientVisitUID               CWhenfirst  ResultValue      Lab
0        2299          3386255  2016-08-15 09:51:54.913        71.70  L1_eGFR
1        2304          3431160  2016-08-30 09:37:00.867        68.90  L1_eGFR
2        2312          5136829  2018-04-28 14:39:32.587        57.60  L1_eGFR
3        2354          7003273  2020-01-09 07:57:21.190        56.99  L1_eGFR
4        2362          3910608  2017-02-20 19:48:50.243        67.60  L1_eGFR
5        2378          3420850  2016-08-26 09:00:24.393        25.90  L1_eGFR
6        2431          3571800  2016-10-18 11:06:38.557        39.00  L1_eGFR
7        2568          4593873  2017-10-19 07:24:03.470        68.50  L1_eGFR
8        2625          4873231  2018-01-27 16:33:43.403        61.90  L1_eGFR
9        2811          5236970  2018-06-01 16:03:47.793        10.70  L1_eGFR

In [12]:
len(group_eGFR1)

22575

In [13]:
#stage G1
print(f'stage G1 : {len(group_eGFR1[(group_eGFR1.ResultValue >= 90)])}')
#stage G2
print(f'stage G2 : {len(group_eGFR1[(group_eGFR1.ResultValue >= 60)&(group_eGFR1.ResultValue < 90)])}')
#stage G3
print(f'stage G3 : {len(group_eGFR1[(group_eGFR1.ResultValue >= 30)&(group_eGFR1.ResultValue < 60)])}')
#stage G4
print(f'stage G4 : {len(group_eGFR1[(group_eGFR1.ResultValue >= 15)&(group_eGFR1.ResultValue < 30)])}')
#stage G5
print(f'stage G5 : {len(group_eGFR1[(group_eGFR1.ResultValue < 15)])}')

stage G1 : 1091
stage G2 : 7636
stage G3 : 10962
stage G4 : 1445
stage G5 : 1441


    id_PatientUID have eGFR more than and equared 15

In [14]:
id_PatientUID = group_eGFR1[(group_eGFR1.ResultValue >= 15)]
id_PatientUID.head(10)

PatientUID  PatientVisitUID               CWhenfirst  ResultValue      Lab
0         2299          3386255  2016-08-15 09:51:54.913        71.70  L1_eGFR
1         2304          3431160  2016-08-30 09:37:00.867        68.90  L1_eGFR
2         2312          5136829  2018-04-28 14:39:32.587        57.60  L1_eGFR
3         2354          7003273  2020-01-09 07:57:21.190        56.99  L1_eGFR
4         2362          3910608  2017-02-20 19:48:50.243        67.60  L1_eGFR
5         2378          3420850  2016-08-26 09:00:24.393        25.90  L1_eGFR
6         2431          3571800  2016-10-18 11:06:38.557        39.00  L1_eGFR
7         2568          4593873  2017-10-19 07:24:03.470        68.50  L1_eGFR
8         2625          4873231  2018-01-27 16:33:43.403        61.90  L1_eGFR
10        2930          6728115  2019-10-08 23:11:21.057        58.60  L1_eGFR

    Start of accrual : 2016-07-28 16.10.25 433000

In [15]:
df_mintime = id_PatientUID['CWhenfirst'].min()
df_mintime

'2016-07-28 16:10:25.433'

    End of accrual : 2022-12-28 09.3710 007000

In [16]:
df_maxtime = id_PatientUID['CWhenfirst'].max()
df_maxtime

'2022-12-28 09:37:10.007'

    group_eGFR3 is df_eGFR groupby

In [17]:
group_eGFR3 = df_eGFR.groupby('PatientUID')

    End of follow-up (Stage 5) : 2023-04-08 12:22:45 987000

In [18]:
df_stage5 = group_eGFR3.apply(lambda x: (x[x['PatientUID'].isin(id_PatientUID.PatientUID) & (x['ResultValue'] < 15)]).head(1)).reset_index(drop=True)
df_stage5['CWhenfirst'].max()

'2023-04-08 12:22:45.987'

In [19]:
df_stage5

PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
0           2431          9802525  2022-08-30 10:01:39.680        11.58   
1           2568          9773774  2022-08-21 16:37:58.733         9.70   
2           3139         10188494  2022-12-29 07:00:27.410        12.89   
3           3390          4436204  2017-09-02 17:37:46.067        14.30   
4           4196          8220556  2021-03-15 10:08:43.480         7.81   
...          ...              ...                      ...          ...   
2082     3343772         10092829  2022-11-25 08:48:39.057        13.40   
2083     3353309         10553151  2023-04-04 07:36:03.417        13.14   
2084     3364037         10164531  2022-12-18 15:40:21.597         8.91   
2085     3368910          9653384  2022-07-27 06:43:36.680        13.90   
2086     3396748          9883317  2022-09-23 18:24:06.123        14.23   

          Lab  
0     L1_eGFR  
1     L1_eGFR  
2     L1_eGFR  
3     L1_eGFR  
4     L1_eGFR  
...       ...  
2082  L1_eGFR  
2083  L1_eGFR  
2084  L1_eGFR  
2085  L1_eGFR  
2086  L1_eGFR  

[2087 rows x 5 columns]

#Censor

    group_eGFR4 is df_eGFR more than and equared 15

In [20]:
group_eGFR4 = group_eGFR1[(group_eGFR1.ResultValue >= 15)]

In [21]:
group_eGFR4

PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
0            2299          3386255  2016-08-15 09:51:54.913        71.70   
1            2304          3431160  2016-08-30 09:37:00.867        68.90   
2            2312          5136829  2018-04-28 14:39:32.587        57.60   
3            2354          7003273  2020-01-09 07:57:21.190        56.99   
4            2362          3910608  2017-02-20 19:48:50.243        67.60   
...           ...              ...                      ...          ...   
22570     3421578         10151787  2022-12-14 11:30:53.310        47.44   
22571     3426284         10105952  2022-11-29 10:47:37.917        27.88   
22572     3431501         10144649  2022-12-13 13:42:48.510        54.65   
22573     3432056         10154545  2022-12-15 09:48:31.117        70.18   
22574     3437526         10191122  2022-12-26 09:29:11.307        55.94   

           Lab  
0      L1_eGFR  
1      L1_eGFR  
2      L1_eGFR  
3      L1_eGFR  
4      L1_eGFR  
...        ...  
22570  L1_eGFR  
22571  L1_eGFR  
22572  L1_eGFR  
22573  L1_eGFR  
22574  L1_eGFR  

[21134 rows x 5 columns]

In [22]:
map_CWhen = df_stage5.set_index('PatientUID')['CWhenfirst'].to_dict()
map_ResultValue = df_stage5.set_index('PatientUID')['ResultValue'].to_dict()
group_eGFR4['CWhe5'] = group_eGFR4['PatientUID'].map(map_CWhen)
group_eGFR4['ResultValue5'] = group_eGFR4['PatientUID'].map(map_ResultValue)

<ipython-input-22-07072a9e8f9f>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  group_eGFR4['CWhe5'] = group_eGFR4['PatientUID'].map(map_CWhen)
<ipython-input-22-07072a9e8f9f>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  group_eGFR4['ResultValue5'] = group_eGFR4['PatientUID'].map(map_ResultValue)


In [23]:
group_eGFR4.head(10)

PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
0         2299          3386255  2016-08-15 09:51:54.913        71.70   
1         2304          3431160  2016-08-30 09:37:00.867        68.90   
2         2312          5136829  2018-04-28 14:39:32.587        57.60   
3         2354          7003273  2020-01-09 07:57:21.190        56.99   
4         2362          3910608  2017-02-20 19:48:50.243        67.60   
5         2378          3420850  2016-08-26 09:00:24.393        25.90   
6         2431          3571800  2016-10-18 11:06:38.557        39.00   
7         2568          4593873  2017-10-19 07:24:03.470        68.50   
8         2625          4873231  2018-01-27 16:33:43.403        61.90   
10        2930          6728115  2019-10-08 23:11:21.057        58.60   

        Lab                    CWhe5  ResultValue5  
0   L1_eGFR                      NaN           NaN  
1   L1_eGFR                      NaN           NaN  
2   L1_eGFR                      NaN           NaN  
3   L1_eGFR                      NaN           NaN  
4   L1_eGFR                      NaN           NaN  
5   L1_eGFR                      NaN           NaN  
6   L1_eGFR  2022-08-30 10:01:39.680         11.58  
7   L1_eGFR  2022-08-21 16:37:58.733          9.70  
8   L1_eGFR                      NaN           NaN  
10  L1_eGFR                      NaN           NaN

    Censored1
        censored = 1
        failure = 0
    Censored2
        censored right = 1
        lost follow-up = 0



In [24]:
def cenored(ResultValue5):
    return 0 if not pd.isna(ResultValue5) and pd.notna(pd.to_numeric(ResultValue5, errors='coerce')) else 1

group_eGFR4['Censored1'] = group_eGFR4['ResultValue5'].apply(cenored)

<ipython-input-24-006187ab530b>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  group_eGFR4['Censored1'] = group_eGFR4['ResultValue5'].apply(cenored)


In [25]:
group_eGFR4.head(10)

PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
0         2299          3386255  2016-08-15 09:51:54.913        71.70   
1         2304          3431160  2016-08-30 09:37:00.867        68.90   
2         2312          5136829  2018-04-28 14:39:32.587        57.60   
3         2354          7003273  2020-01-09 07:57:21.190        56.99   
4         2362          3910608  2017-02-20 19:48:50.243        67.60   
5         2378          3420850  2016-08-26 09:00:24.393        25.90   
6         2431          3571800  2016-10-18 11:06:38.557        39.00   
7         2568          4593873  2017-10-19 07:24:03.470        68.50   
8         2625          4873231  2018-01-27 16:33:43.403        61.90   
10        2930          6728115  2019-10-08 23:11:21.057        58.60   

        Lab                    CWhe5  ResultValue5  Censored1  
0   L1_eGFR                      NaN           NaN          1  
1   L1_eGFR                      NaN           NaN          1  
2   L1_eGFR                      NaN           NaN          1  
3   L1_eGFR                      NaN           NaN          1  
4   L1_eGFR                      NaN           NaN          1  
5   L1_eGFR                      NaN           NaN          1  
6   L1_eGFR  2022-08-30 10:01:39.680         11.58          0  
7   L1_eGFR  2022-08-21 16:37:58.733          9.70          0  
8   L1_eGFR                      NaN           NaN          1  
10  L1_eGFR                      NaN           NaN          1

In [26]:
group_eGFR4['Censored2'] = (group_eGFR4.Censored1 == 1) & (group_eGFR2.CWhenfirst >= df_stage5['CWhenfirst'].max())
group_eGFR4['Censored2'] = group_eGFR4['Censored2'].astype(int)
group_eGFR4.head(10)

<ipython-input-26-727b82b8702d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  group_eGFR4['Censored2'] = (group_eGFR4.Censored1 == 1) & (group_eGFR2.CWhenfirst >= df_stage5['CWhenfirst'].max())
<ipython-input-26-727b82b8702d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  group_eGFR4['Censored2'] = group_eGFR4['Censored2'].astype(int)


PatientUID  PatientVisitUID               CWhenfirst  ResultValue  \
0         2299          3386255  2016-08-15 09:51:54.913        71.70   
1         2304          3431160  2016-08-30 09:37:00.867        68.90   
2         2312          5136829  2018-04-28 14:39:32.587        57.60   
3         2354          7003273  2020-01-09 07:57:21.190        56.99   
4         2362          3910608  2017-02-20 19:48:50.243        67.60   
5         2378          3420850  2016-08-26 09:00:24.393        25.90   
6         2431          3571800  2016-10-18 11:06:38.557        39.00   
7         2568          4593873  2017-10-19 07:24:03.470        68.50   
8         2625          4873231  2018-01-27 16:33:43.403        61.90   
10        2930          6728115  2019-10-08 23:11:21.057        58.60   

        Lab                    CWhe5  ResultValue5  Censored1  Censored2  
0   L1_eGFR                      NaN           NaN          1          0  
1   L1_eGFR                      NaN           NaN          1          0  
2   L1_eGFR                      NaN           NaN          1          0  
3   L1_eGFR                      NaN           NaN          1          0  
4   L1_eGFR                      NaN           NaN          1          0  
5   L1_eGFR                      NaN           NaN          1          0  
6   L1_eGFR  2022-08-30 10:01:39.680         11.58          0          0  
7   L1_eGFR  2022-08-21 16:37:58.733          9.70          0          0  
8   L1_eGFR                      NaN           NaN          1          0  
10  L1_eGFR                      NaN           NaN          1          0

In [27]:
group_eGFR4.to_csv('/content/drive/My Drive/ProjectCKD/group_eGFR4.csv', index=False)